windows: ungraceful mycelium stop when myceliumflut UI/window closed

Steps to reproduce:
- start myceliumflut using `flutter run`
- click `start mycelium` button
- close myceliumflut (without clicking `stop mycelium` button first)

notes: there is no issue if we click `stop mycelium` button first before close the `mycelumflut`

expected:
- `flutter run` command finished
- `ping` to myceliumflut become failed

what happens:
- `ping` still OK, it means mycelium not killed
- `flutter run` command hanging
- 

What i got in the console
![image](https://github.com/user-attachments/assets/b8894202-1314-47cc-b9c8-2ef41567b357)
- subsequent `flutter run` also failed with below error, restart Windows make it work again
```console
PS C:\Users\Administrator\baru\myceliumflut> flutter run
Launching lib\main.dart on Windows in debug mode...
LINK : fatal error LNK1168: cannot open C:\Users\Administrator\baru\myceliumflut\build\windows\x64\runner\Debug\myceliumflut.exe for writing [C:\Users\Administrator\baru\myceliumflut\build\windows\x64\runner\myceliumflut.vcxproj]
Building Windows application...                                    23.0s
Error: Build process failed.
```


From the log, the `terminate` handler added at #71 (which supposed to work on Windows) doesn't work on Windows

--- a/lib/main.dart
+++ b/lib/main.dart
@@ -8,6 +8,7 @@ import 'package:path_provider/path_provider.dart';
 import 'package:logging/logging.dart';
 
 import 'package:flutter_desktop_sleep/flutter_desktop_sleep.dart';
+import 'package:flutter_window_close/flutter_window_close.dart';
 import 'dart:ffi' as ffi;
 
 import 'package:ffi/ffi.dart';
@@ -67,8 +68,7 @@ class _MyAppState extends State<MyApp> {
     // flutter desktop sleep plugin only works on macos and windows
     _flutterDesktopSleepPlugin.setWindowSleepHandler((String? s) async {
       if (s != null) {
-        if (s == 'sleep') {
-        } else if (s == 'woke_up') {
+        if (s == 'woke_up') {
           if (_isStarted) {
             _logger.info("[wake up handler]stopping mycelium");
             stopMycelium();
@@ -81,14 +81,33 @@ class _MyAppState extends State<MyApp> {
             startMycelium();
           }
         } else if (s == 'terminate_app') {
-          _logger.info("TERMINATE_APP");
-          if (_isStarted) {
-            stopMycelium();
+          // only for macos because:
+          // 1. it supposed to work on Windows as well, but it doesn't
+          // 2. to make it clear that it's only for macos
+          if (Platform.isMacOS) {
+            _logger.info("TERMINATE_APP");
+            if (_isStarted) {
+              stopMycelium();
+            }
+            _flutterDesktopSleepPlugin.terminateApp();
           }
-          _flutterDesktopSleepPlugin.terminateApp();
+        } else {
+          _logger
+              .info("[flutter_desktop_sleep handler]Unknown event handler: $s");
         }
       }
     });
+
+    // only for windows because macos already has own handler
+    if (Platform.isWindows) {
+      FlutterWindowClose.setWindowShouldCloseHandler(() async {
+        _logger.info("Window close handler");
+        if (_isStarted) {
+          stopMycelium();
+        }
+        return true;
+      });
+    }
   }
 
   void methodHandler(String methodName) {

--- a/macos/Flutter/GeneratedPluginRegistrant.swift
+++ b/macos/Flutter/GeneratedPluginRegistrant.swift
@@ -6,9 +6,11 @@ import FlutterMacOS
 import Foundation
 
 import flutter_desktop_sleep
+import flutter_window_close
 import path_provider_foundation
 
 func RegisterGeneratedPlugins(registry: FlutterPluginRegistry) {
   FlutterDesktopSleepPlugin.register(with: registry.registrar(forPlugin: "FlutterDesktopSleepPlugin"))
+  FlutterWindowClosePlugin.register(with: registry.registrar(forPlugin: "FlutterWindowClosePlugin"))
   PathProviderPlugin.register(with: registry.registrar(forPlugin: "PathProviderPlugin"))
 }

--- a/pubspec.yaml
+++ b/pubspec.yaml
@@ -41,6 +41,7 @@ dependencies:
   flutter_desktop_sleep: ^0.0.2
   path: 1.9.0
   ffi: 2.1.2
+  flutter_window_close: ^1.2.0
 
 dev_dependencies:
   flutter_test:

--- a/windows/flutter/generated_plugin_registrant.cc
+++ b/windows/flutter/generated_plugin_registrant.cc
@@ -7,8 +7,11 @@
 #include "generated_plugin_registrant.h"
 
 #include <flutter_desktop_sleep/flutter_desktop_sleep_plugin_c_api.h>
+#include <flutter_window_close/flutter_window_close_plugin.h>
 
 void RegisterPlugins(flutter::PluginRegistry* registry) {
   FlutterDesktopSleepPluginCApiRegisterWithRegistrar(
       registry->GetRegistrarForPlugin("FlutterDesktopSleepPluginCApi"));
+  FlutterWindowClosePluginRegisterWithRegistrar(
+      registry->GetRegistrarForPlugin("FlutterWindowClosePlugin"));
 }

--- a/windows/flutter/generated_plugins.cmake
+++ b/windows/flutter/generated_plugins.cmake
@@ -4,6 +4,7 @@
 
 list(APPEND FLUTTER_PLUGIN_LIST
   flutter_desktop_sleep
+  flutter_window_close
 )
 
 list(APPEND FLUTTER_FFI_PLUGIN_LIST

